libcnb/examples/example-02-ruby-sample/src/main.rs

```diff
@@ -1,22 +1,21 @@
-use crate::layers::BundlerLayerLifecycle;
-use crate::layers::RubyLayerLifecycle;
+use crate::layers::{BundlerLayer, RubyLayer};
 use libcnb::build::{BuildContext, BuildResult, BuildResultBuilder};
-use libcnb::buildpack_main;
 use libcnb::data::launch::{Launch, Process};
 use libcnb::detect::{DetectContext, DetectResult, DetectResultBuilder};
 use libcnb::generic::GenericPlatform;
-use libcnb::layer_lifecycle::execute_layer_lifecycle;
-use libcnb::Buildpack;
+use libcnb::layer_env::TargetLifecycle;
+use libcnb::{buildpack_main, Buildpack, Env};
+
 use serde::Deserialize;
 
 mod layers;
 
 #[derive(Deserialize, Debug)]
-struct RubyBuildpackMetadata {
+pub struct RubyBuildpackMetadata {
     pub ruby_url: String,
 }
 
-struct RubyBuildpack;
+pub struct RubyBuildpack;
```

Comment:
We use `RubyBuildpack` (which does contain a reference to the `RubyBuildpackMetadata` type) in the modules for the layers of this buildpack.

I just checked, making them private would yield a compile error as expected, surfacing the above:
```
error[E0446]: private type `RubyBuildpack` in public interface
  --> libcnb/examples/example-02-ruby-sample/src/layers/bundler.rs:26:5
   |
26 |     type Buildpack = RubyBuildpack;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't leak private type
   |
  ::: libcnb/examples/example-02-ruby-sample/src/main.rs:18:1
   |
18 | struct RubyBuildpack;
   | --------------------- `RubyBuildpack` declared as private

error[E0446]: private type `RubyBuildpack` in public interface
  --> libcnb/examples/example-02-ruby-sample/src/layers/ruby.rs:20:5
   |
20 |     type Buildpack = RubyBuildpack;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't leak private type
   |
  ::: libcnb/examples/example-02-ruby-sample/src/main.rs:18:1
   |
18 | struct RubyBuildpack;
   | --------------------- `RubyBuildpack` declared as private

```

Reply:
Ah I didn't know that referencing parent modules would require them to be public. Is that because the associated type then gets used by libcnb, so outside the Ruby example crate?

Reply:
You can access private functions and types from a modules parent, but not the other way around:

```rust
mod a {
    fn foo() {}

    fn bar() {
        // Not allowed
        b::baz();
    }

    mod b {
        fn baz() {
            // Allowed
            super::foo();
        }
    }
}
```

Reply:
Yeah I realise, which is why I was surprised. This usage is accessing the parent.

Reply:
Ah I missed this in the error message:
> can't leak private type

So I was right, it's because of the associated type.

Reply:
Ah so https://doc.rust-lang.org/error-index.html#E0446 says:

> There are two ways to solve this error. The first is to make the public type signature only public to a module that also has access to the private type.

(And agree this isn't that important, I was mainly curious)

Reply:
So with this it works:

```rust
pub(crate) struct RubyLayer;
```

But I'm not fussed about changing this either way.